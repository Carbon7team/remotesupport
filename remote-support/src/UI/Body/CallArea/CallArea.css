#wrapper-callarea {
    background-color: aliceblue;
    padding: 5em;
}