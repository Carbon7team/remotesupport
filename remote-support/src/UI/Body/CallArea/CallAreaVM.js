class CallAreaVM {}

export default CallAreaVM;