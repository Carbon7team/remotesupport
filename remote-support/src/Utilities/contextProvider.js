import {createContext, useContext} from "react";
import Rootstore from "../ModelStores/RootStore";

export const AppContext = createContext(Rootstore);
export const AppContextProvider = AppContext.Provider;
export const useStore = () => useContext(AppContext);