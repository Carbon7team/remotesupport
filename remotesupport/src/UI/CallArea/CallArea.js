import React from 'react';
import { Typography, AppBar } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';

import VideoPlayer from './components/VideoPlayer';
import Sidebar from './components/Sidebar';
import Notifications from './components/Notifications';

const useStyles = makeStyles((theme) => ({
  image: {
    marginLeft: '15px',
  },
  wrapper: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'right',
    width: '100%',
  },
}));

const CallArea = (props) => {

  const classes = useStyles();
  
  const isLogged = props.isLogged;

  return (
    <>
    {!isLogged && (
      <div id='message-not-logged-wrapper'>
      <h2>SOCOMEC SUPPORT</h2>
      <p>Effettua il login</p>
      <p>All'avvio il tuo stato sar√† impostato come <span>Disponibile</span></p>
      </div>
    )}
    { isLogged && (
    <div className={classes.wrapper}>
      <h2>Video Chat</h2>
      <VideoPlayer/>
    </div>)
    }
    </>
  );
};

export default CallArea;
