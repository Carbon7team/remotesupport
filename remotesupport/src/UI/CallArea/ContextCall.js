import React, { createContext, useState, useRef, useEffect, Children } from 'react';
import socketIOClient from 'socket.io-client';
import Peer from 'peerjs';
const ENDPOINT = "http://localhost:4000"; // server address

const ContextCall = createContext();

// const socket = io('http://localhost:5000');
// const socket = io('');

const ContextProvider = ({children}) => {
  const [callAccepted, setCallAccepted] = useState(false);
  const [callEnded, setCallEnded] = useState(true);
  const [stream, setStream] = useState(undefined);
  const [nameUser, setNameUser] = useState('');
  const [surnameUser, setSurnameUser] = useState('');
  const [companyUser, setCompanyUser] = useState('');
  const [idUser, setIdUser] = useState('');
  const [jsonData, setJsonData] = useState(undefined);

  const myVideo = useRef(null);
  const userVideo = useRef(null);
  const peerInstance = useRef(null);

  useEffect(() => {

    const socket = socketIOClient(ENDPOINT);
    
    socket.send({"identificazione" : "tecnico","tipo": "registrazione","idTecnico": "tecnico1"}); // inserire username di accesso tecnico
    
    socket.on("message", data => {
      if(data.identificazione == "virtualDisplay"){
        if(data.tipo == "chiamata"){
          setIdUser(data.idvirtualDisplay);
        }
      }
    });
  }, []);
  
  const sendCall = (idUser) => {

    const peer = new Peer("tecnico1" /* var con username del tecnico */, {
      host: 'localhost',
      port: 9000,
      path: '/myapp'
    });

    setCallAccepted(true);
    setCallEnded(false);

    peerInstance.current = peer;

    navigator.mediaDevices.getUserMedia({ video: true, audio: true }).then((mediaStream) => {
      
      setStream(mediaStream);

      myVideo.current.srcObject = mediaStream;
      myVideo.current.play();

      const call = peerInstance.current.call(idUser, mediaStream)

      call.on('stream', (remoteStream) => {
        userVideo.current.srcObject = remoteStream
        userVideo.current.play();
      });
    });

    const connection = peer.connect(idUser);

    connection.on('open', () => {
      connection.on('data', function(data) {
        setJsonData(JSON.stringify(data)); // setting string JSON to stamp later
      });
    });

}

  const toggleAudio = () => {
    stream.getAudioTracks().forEach(track => track.enabled = !track.enabled);
  }

  const toggleVideo = () => {
    stream.getVideoTracks().forEach(track => track.enabled = !track.enabled);
  }

  const leaveCall = () => {

    setCallEnded(true);
    setCallAccepted(false);
    setStream(undefined);
    setJsonData(undefined);

    peerInstance.current.destroy();

  }

  

  return(
    <ContextCall.Provider value={{
      callAccepted,
      callEnded,
      idUser,
      stream,
      myVideo,
      userVideo,
      jsonData,
      toggleAudio,
      toggleVideo,
      sendCall,
      leaveCall
    }} >
      {children}
    </ContextCall.Provider>
  )

};

export {ContextProvider, ContextCall};