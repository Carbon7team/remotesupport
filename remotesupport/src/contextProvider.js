import {createContext, useContext} from "react";
import Dataset from './store'

export const AppContext = createContext(Dataset);
export const AppContextProvider = AppContext.Provider;
export const useStore = () => useContext(AppContext);