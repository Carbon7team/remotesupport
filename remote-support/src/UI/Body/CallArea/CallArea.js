import { observer } from "mobx-react-lite";
import React, { useState, useRef, useEffect } from "react";
import Sidebar from "./SideBar/Sidebar";
import "./CallArea.css";
import { Button } from "@material-ui/core";
import { useStore } from "../../../Utilities/contextProvider";
import Peer from "peerjs";
import socketIOClient from "socket.io-client";
const ENDPOINT = "http://localhost:4000";

const CallArea = observer(() => {
  const rootstore = useStore();
  const peerTech = useRef(null);
  const call = useRef(null);
  const connectionDataChannel = useRef(null);
  const audioTech = useRef(null);
  const audioUser = useRef(null);
  const [idUserClient, setIdUserClient] = useState(null);

  const socket = socketIOClient(ENDPOINT);

  const { callAccepted, callEnded, requestReceived } = rootstore.stateUIStore;

  const sendCall = (idUserClient) => {
    const peer = new Peer(rootstore.stateUIStore.idTech, {
      host: "localhost",
      port: 9000,
      path: "/myapp",
    });

    rootstore.stateUIStore.setCallAccepted(true);
    rootstore.stateUIStore.setCallEnded(false);

    peerTech.current = peer;

    navigator.mediaDevices
      .getUserMedia({ video: false, audio: true })
      .then((mediaStream) => {
        rootstore.stateUIStore.setStreamTech(mediaStream);

        audioTech.current.srcObject = mediaStream;

        call.current = peerTech.current.call(idUserClient, mediaStream);

        call.current.on("stream", (remoteStream) => {
          audioUser.current.srcObject = remoteStream;
        });
      });

    connectionDataChannel.current = peerTech.current.connect(idUserClient);

    connectionDataChannel.current.on("open", () => {
      connectionDataChannel.current.on("data", function (data) {
        try {
          rootstore.datasetStore.statesFromJSON(JSON.parse(data).status);
          rootstore.datasetStore.alarmsFromJSON(JSON.parse(data).alarms);
          rootstore.datasetStore.measurementsFromJSON(
            JSON.parse(data).measurements
          );
        } catch {
          rootstore.datasetStore.resetStates();
          rootstore.datasetStore.resetAlarms();
          rootstore.datasetStore.resetMeasurements();
        }
      });
    });
  };

  async function availabilityFetch(valueAvailability) {
    const data = {
      disponibility: valueAvailability,
      user_id: rootstore.stateUIStore.idTech,
    };
    console.log("availability :" + valueAvailability);
    return fetch("http://localhost:4000/changeDisponibility", {
      method: "POST",
      headers: {
        Authentication: rootstore.stateUIStore.tokenAuth,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
    })
      .then((data) => data) //print data to console
      .catch((err) => console.log("Request Failed", err)); // Catch errors
  }

  const declineCall = async () => {
    const data_to_send = {
      type: "refuse",
      user_id: idUserClient,
    };

    const fetchAvaiability = await availabilityFetch(true);
    if (fetchAvaiability.status == 200) {
      rootstore.stateUIStore.setRequestReceived(false);
    }

    socket.send(data_to_send);
    rootstore.stateUIStore.setNameClient(null);
    rootstore.stateUIStore.setSurnameClient(null);
    rootstore.stateUIStore.setEmailClient(null);
    rootstore.stateUIStore.setUsernameClient(null);
    setIdUserClient(null);
  };

  const leaveCall = async () => {
    console.log(call);
    console.log(connectionDataChannel);
    console.log(peerTech);
    console.log(call.current.open);
    await availabilityFetch(true);
    rootstore.stateUIStore.setCallEnded(true);
    rootstore.stateUIStore.setCallAccepted(false);
    rootstore.stateUIStore.setStreamTech(undefined);
    rootstore.stateUIStore.setRequestReceived(false);
    rootstore.datasetStore.resetStates();
    rootstore.datasetStore.resetAlarms();
    rootstore.datasetStore.resetMeasurements();
    setIdUserClient(null);
    audioTech.current = null;
    audioUser.current = null;

    // come qui https://github.com/adrianhajdin/project_video_chat/blob/master/client/src/Context.js line 78
    call.current.close();
    connectionDataChannel.current.close();
    peerTech.current.destroy();
    console.log(call.current.open);
    console.log(call);
    console.log(connectionDataChannel);
    console.log(peerTech);
  };

  const setDataRequestClient = () => {
    const data_registration = {
      type: "registration",
      idUser: rootstore.stateUIStore.idTech,
    };

    socket.send(data_registration);

    socket.on("message", async (data) => {
      if (data.type === "call") {
        const fetchAvaiability = await availabilityFetch(false);
        if (fetchAvaiability.status == 200) {
          rootstore.stateUIStore.setRequestReceived(true);
        }

        rootstore.stateUIStore.setNameClient(data.first_name);
        rootstore.stateUIStore.setSurnameClient(data.last_name);
        rootstore.stateUIStore.setEmailClient(data.email);
        rootstore.stateUIStore.setUsernameClient(data.username);
        setIdUserClient(data.user_id);
      }
    });
  };

  useEffect(() => {
    if (!requestReceived) {
      setDataRequestClient();
    }
  }, []);

  return (
    <div id="wrapper-callarea">
      <h2>Call Area</h2>

      {!callAccepted && callEnded && requestReceived && (
        <>
          <p>Do you accept this request?</p>
          <Button
            id="accept-button"
            onClick={() => {
              sendCall(idUserClient);
            }}
          >
            Accept
          </Button>
          <Button
            id="decline-button"
            onClick={() => {
              declineCall();
            }}
          >
            Decline
          </Button>
        </>
      )}

      {callAccepted && !callEnded && (
        <>
          <p>On call</p>
          {/* {peerTech.current.pingInterval < 100 && <p>Good connection</p>}
          {peerTech.current.pingInterval > 100 &&
            peerTech.current.pingInterval < 300 && <p>Medium connection</p>}
          {peerTech.current.pingInterval > 300 && <p>Bad connection</p>} */}
          <audio ref={audioTech} />
          <Sidebar leaveCall={leaveCall} />
          <audio ref={audioUser} />
        </>
      )}
    </div>
  );
});

export default CallArea;
