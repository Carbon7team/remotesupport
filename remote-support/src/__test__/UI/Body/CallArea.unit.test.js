import { render } from "@testing-library/react";
import { create } from "react-test-renderer";
import Peer from "peerjs";
import CallArea from "../../../UI/Body/CallArea/CallArea";
import { useStore } from "../../../Utilities/contextProvider";

jest.mock("../../../Utilities/contextProvider");

describe("CallArea correct render", () => {
  const vars = jest.fn().mockReturnValue({
    peerTech: { current: new Peer() },
    call: { current: null },
    connectionDataChannel: { current: null },
  });

  test("CallArea render not in call", () => {
    useStore.mockReturnValue({
      stateUIStore: {
        callAccepted: false,
        callEnded: true,
      },
    });

    let tree = create(<CallArea vars={vars} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("CallArea render in call", () => {
    useStore.mockReturnValue({
      stateUIStore: {
        callAccepted: true,
        callEnded: false,
        streamTech: null,
      },
    });

    render(<CallArea vars={vars} />);
    expect(document.getElementById("wrapper-sidebar")).toBeInTheDocument();
  });
});
