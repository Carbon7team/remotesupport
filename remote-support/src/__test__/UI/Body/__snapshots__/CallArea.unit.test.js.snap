// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallArea correct render CallArea render not in call 1`] = `
<div
  id="wrapper-callarea"
>
  <h2>
    Call Area
  </h2>
</div>
`;
